Support google_compute_instance import

The initial idea was to try to update the configuration for google compute disks:
`disk` ==> `boot_disk`/`attached_disk` as required in version 10  to prevent all the deprecation errors, however this action involves destroying disk and the instance and then re-provision them. 

To avoid re-provisioning I thought we can manually remove all the resources from the state and re-import them on by one with the new configuration. Unfortunately importing `google_compute_instance` is currently not supported. I think we will stay with deprecation warnings and wait until this issue is resolve to update our config . 


### Terraform Version
Terraform v0.10.2


### Affected Resource(s)
- google_compute_instance

### Terraform Configuration Files
```hcl
resource "google_compute_instance" "compute-recommendation" {
  count = "${var.gce_recommendation_compute_count}"
  name = "${var.environment}-${var.client}-compute-recommendation-${count.index}"
  machine_type = "${var.gce_recommendation_compute_machine_type}"
  zone = "${var.gce_region}-b"
  tags = ["compute-recommendation"]

  boot_disk {
    source = "${element(google_compute_disk.compute-recommendation-disk.*.name, count.index)}"
  }
  network_interface {
    network = "${var.environment}"
    access_config {
      nat_ip =  "${element(google_compute_address.compute-recommendation-static-ips.*.address, count.index)}"
    }
  }
  metadata {
    sshKeys = "${var.gce_ssh_user}:${file("${var.gce_ssh_pub_key_file}")}"
  }
  service_account {
    scopes = "${var.service_account_scopes}"
  }
}
```

### Debug Output
https://gist.github.com/nikoren/887544482ab7597789b36b08a021a3af

### Panic Output


### Expected Behavior
Compute instance should be imported

### Actual Behavior
Get error 

```
Error importing: 1 error(s) occurred:

* module.dev-tst-environment.google_compute_instance.compute-recommendation (import id: 49023364572610319): import module.dev-tst-environment.google_compute_instance.compute-recommendation (id: 49023364572610319): resource google_compute_instance doesn't support import
```
### Steps to Reproduce
1. create `gce_compute_instance` 
2. remove instance from state `terraform state rm gce_compute_instance.name`
3. Try to import the instance `terraform import google_compute_instance.compute-recommendation ${instance_id}`

I guess this open issue is related to what I am noticing as well, 

I am on 

$ terraform --version
Terraform v0.11.1
+ provider.google v1.1.0

and trying to do,

$ terraform import google_compute_instance.ac01 ac01
google_compute_instance.ac01: Importing from ID "ac01"...
Error importing: 1 error(s) occurred:

* google_compute_instance.ac01 (import id: ac01): import google_compute_instance.ac01 (id: ac01): resource google_compute_instance doesn't support import


Logs...
2017/12/17 11:36:08 [ERROR] root: eval: *terraform.EvalImportState, err: import google_compute_instance.ac01 (id: ac01): resource google_compute_instance doesn't support import
2017/12/17 11:36:08 [ERROR] root: eval: *terraform.EvalSequence, err: import google_compute_instance.ac01 (id: ac01): resource google_compute_instance doesn't support import

--- a/google/resource_compute_instance.go
+++ b/google/resource_compute_instance.go
@@ -26,6 +26,9 @@ func resourceComputeInstance() *schema.Resource {
 		Read:   resourceComputeInstanceRead,
 		Update: resourceComputeInstanceUpdate,
 		Delete: resourceComputeInstanceDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceComputeInstanceImportState,
+		},
 
 		SchemaVersion: 6,
 		MigrateState:  resourceComputeInstanceMigrateState,
@@ -738,12 +741,10 @@ func resourceComputeInstanceRead(d *schema.ResourceData, meta interface{}) error
 
 	md := flattenMetadataBeta(instance.Metadata)
 
-	if _, scriptExists := d.GetOk("metadata_startup_script"); scriptExists {
-		d.Set("metadata_startup_script", md["startup-script"])
-		// Note that here we delete startup-script from our metadata list. This is to prevent storing the startup-script
-		// as a value in the metadata since the config specifically tracks it under 'metadata_startup_script'
-		delete(md, "startup-script")
-	}
+	d.Set("metadata_startup_script", md["startup-script"])
+	// Note that here we delete startup-script from our metadata list. This is to prevent storing the startup-script
+	// as a value in the metadata since the config specifically tracks it under 'metadata_startup_script'
+	delete(md, "startup-script")
 
 	existingMetadata := d.Get("metadata").(map[string]interface{})
 
@@ -794,6 +795,7 @@ func resourceComputeInstanceRead(d *schema.ResourceData, meta interface{}) error
 	// Set the tags fingerprint if there is one.
 	if instance.Tags != nil {
 		d.Set("tags_fingerprint", instance.Tags.Fingerprint)
+		d.Set("tags", convertStringArrToInterface(instance.Tags.Items))
 	}
 
 	if len(instance.Labels) > 0 {
@@ -859,6 +861,7 @@ func resourceComputeInstanceRead(d *schema.ResourceData, meta interface{}) error
 	d.Set("self_link", ConvertSelfLinkToV1(instance.SelfLink))
 	d.Set("instance_id", fmt.Sprintf("%d", instance.Id))
 	d.Set("project", project)
+	d.Set("name", instance.Name)
 	d.SetId(instance.Name)
 
 	return nil
@@ -1237,6 +1240,20 @@ func resourceComputeInstanceDelete(d *schema.ResourceData, meta interface{}) err
 
 	d.SetId("")
 	return nil
+}
+
+func resourceComputeInstanceImportState(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	parts := strings.Split(d.Id(), "/")
+
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("Invalid import id %q. Expecting {project}/{zone}/{instance_name}", d.Id())
+	}
+
+	d.Set("project", parts[0])
+	d.Set("zone", parts[1])
+	d.SetId(parts[2])
+
+	return []*schema.ResourceData{d}, nil
 }
 
 func expandBootDisk(d *schema.ResourceData, config *Config, zone *compute.Zone, project string) (*computeBeta.AttachedDisk, error) {

--- a/google/resource_compute_instance_test.go
+++ b/google/resource_compute_instance_test.go
@@ -37,6 +37,12 @@ func TestAccComputeInstance_basic1(t *testing.T) {
 					testAccCheckComputeInstanceDisk(&instance, instanceName, true, true),
 				),
 			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params", "create_timeout"},
+			},
 		},
 	})
 }
@@ -246,6 +252,12 @@ func TestAccComputeInstance_attachedDisk(t *testing.T) {
 					testAccCheckComputeInstanceDisk(&instance, diskName, false, false),
 				),
 			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
+			},
 		},
 	})
 }
@@ -269,6 +281,12 @@ func TestAccComputeInstance_attachedDisk_sourceUrl(t *testing.T) {
 						"google_compute_instance.foobar", &instance),
 					testAccCheckComputeInstanceDisk(&instance, diskName, false, false),
 				),
+			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
 			},
 		},
 	})
@@ -347,6 +365,11 @@ func TestAccComputeInstance_bootDisk_source(t *testing.T) {
 					testAccCheckComputeInstanceBootDisk(&instance, diskName),
 				),
 			},
+			resource.TestStep{
+				ResourceName:  "google_compute_instance.foobar",
+				ImportState:   true,
+				ImportStateId: fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+			},
 		},
 	})
 }
@@ -370,6 +393,12 @@ func TestAccComputeInstance_bootDisk_sourceUrl(t *testing.T) {
 						"google_compute_instance.foobar", &instance),
 					testAccCheckComputeInstanceBootDisk(&instance, diskName),
 				),
+			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
 			},
 		},
 	})
@@ -417,6 +446,12 @@ func TestAccComputeInstance_scratchDisk(t *testing.T) {
 						"google_compute_instance.scratch", &instance),
 					testAccCheckComputeInstanceScratchDisk(&instance, []string{"NVME", "SCSI"}),
 				),
+			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
 			},
 		},
 	})
@@ -511,6 +546,12 @@ func TestAccComputeInstance_service_account(t *testing.T) {
 						"https://www.googleapis.com/auth/userinfo.email"),
 				),
 			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
+			},
 		},
 	})
 }
@@ -532,6 +573,12 @@ func TestAccComputeInstance_scheduling(t *testing.T) {
 					testAccCheckComputeInstanceExists(
 						"google_compute_instance.foobar", &instance),
 				),
+			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
 			},
 		},
 	})
@@ -556,6 +603,12 @@ func TestAccComputeInstance_subnet_auto(t *testing.T) {
 					testAccCheckComputeInstanceHasSubnet(&instance),
 				),
 			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
+			},
 		},
 	})
 }
@@ -578,6 +631,12 @@ func TestAccComputeInstance_subnet_custom(t *testing.T) {
 						"google_compute_instance.foobar", &instance),
 					testAccCheckComputeInstanceHasSubnet(&instance),
 				),
+			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
 			},
 		},
 	})
@@ -700,6 +759,12 @@ func TestAccComputeInstance_forceChangeMachineTypeManually(t *testing.T) {
 				),
 				ExpectNonEmptyPlan: true,
 			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params", "create_timeout"},
+			},
 		},
 	})
 }
@@ -724,6 +789,12 @@ func TestAccComputeInstance_multiNic(t *testing.T) {
 					testAccCheckComputeInstanceHasMultiNic(&instance),
 				),
 			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-central1-a", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
+			},
 		},
 	})
 }
@@ -745,6 +816,12 @@ func TestAccComputeInstance_guestAccelerator(t *testing.T) {
 					testAccCheckComputeInstanceExists("google_compute_instance.foobar", &instance),
 					testAccCheckComputeInstanceHasGuestAccelerator(&instance, "nvidia-tesla-k80", 1),
 				),
+			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-east1-d", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
 			},
 		},
 	})
@@ -769,6 +846,12 @@ func TestAccComputeInstance_minCpuPlatform(t *testing.T) {
 					testAccCheckComputeInstanceHasMinCpuPlatform(&instance, "Intel Haswell"),
 				),
 			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-east1-d", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
+			},
 		},
 	})
 }
@@ -791,6 +874,12 @@ func TestAccComputeInstance_primaryAliasIpRange(t *testing.T) {
 					testAccCheckComputeInstanceHasAliasIpRange(&instance, "", "/24"),
 				),
 			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-east1-d", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
+			},
 		},
 	})
 }
@@ -812,6 +901,12 @@ func TestAccComputeInstance_secondaryAliasIpRange(t *testing.T) {
 					testAccCheckComputeInstanceExists("google_compute_instance.foobar", &instance),
 					testAccCheckComputeInstanceHasAliasIpRange(&instance, "inst-test-secondary", "172.16.0.0/24"),
 				),
+			},
+			resource.TestStep{
+				ResourceName:            "google_compute_instance.foobar",
+				ImportState:             true,
+				ImportStateId:           fmt.Sprintf("%s/%s/%s", getTestProjectFromEnv(), "us-east1-d", instanceName),
+				ImportStateVerifyIgnore: []string{"boot_disk.0.initialize_params"},
 			},
 		},
 	})

--- a/website/docs/r/compute_instance.html.markdown
+++ b/website/docs/r/compute_instance.html.markdown
@@ -275,7 +275,7 @@ exported:
 
 ## Import
 
-~> **Note:** The fields `boot_disk.0.initialize_params`, `boot_disk.0.disk_entryption_raw` and `attached_disk.*.disk_encryption_key_raw` cannot be imported automatically. The API doesn't return this information. You will need to update your state manually for this field after importing the resource.
+~> **Note:** The fields `boot_disk.0.initialize_params`, `boot_disk.0.disk_entryption_raw` and `attached_disk.*.disk_encryption_key_raw` cannot be imported automatically. The API doesn't return this information. If you are setting one of these fields in your config, you will need to update your state manually after importing the resource.
 
 Instances can be imported using the `project`, `zone` and `name`, e.g.
 

--- a/website/docs/r/compute_instance.html.markdown
+++ b/website/docs/r/compute_instance.html.markdown
@@ -272,3 +272,13 @@ exported:
 * `disk.0.disk_encryption_key_sha256` - The [RFC 4648 base64](https://tools.ietf.org/html/rfc4648#section-4)
     encoded SHA-256 hash of the [customer-supplied encryption key]
     (https://cloud.google.com/compute/docs/disks/customer-supplied-encryption) that protects this resource.
+
+## Import
+
+~> **Note:** The fields `boot_disk.0.initialize_params`, `boot_disk.0.disk_entryption_raw` and `attached_disk.*.disk_encryption_key_raw` cannot be imported automatically. The API doesn't return this information. You will need to update your state manually for this field after importing the resource.
+
+Instances can be imported using the `project`, `zone` and `name`, e.g.
+
+```
+$ terraform import google_compute_instance.default gcp-project/us-central1-a/test
+```

